pre-commit

```diff
@@ -18,22 +16,62 @@ declare -a descriptions=(
     "AWS account number"
     "16 byte hex (e.g. S3 bucket name)"
     "NINO"
-    "IP address"
     "Private key (e.g. rsa private key, openssh private key)"
-    "Email addresses"
     "Regions embedded as part of resource descriptions"
 )
 
+declare -a email_patterns=(
+    "\b[A-Za-z0-9._%+-]{1,}@[A-Za-z0-9]{1,}.[A-Za-z0-9]{1,4}.[A-Za-z0-9]{1,4}\b"
+)
+
+declare -a email_descriptions=(
+    "Email addresses"
+)
+
+declare -a ip_patterns=(
+    "\b(?:[0-9]{1,3}\.){3}[0-9]{1,3}\b"
+)
+
+declare -a ip_descriptions=(
+    "IP address"
+)
+
 if [ -d ".gitsecret" ]; then
     git secret hide
 fi
 
+
 match=0
 for i in "${!patterns[@]}"
 do
     git diff-index -p -M --cached HEAD -- | grep -v "Subproject commit" |
     grep '^+[^+]' | grep -Eq "${patterns[$i]}" &&
     echo "Blocking commit: ${descriptions[$i]} detected in patch" &&
+    echo "This part of your change triggered this issue:" &&
+    git diff-index -p -M --cached HEAD -- | grep -v "Subproject commit" |
+    grep '^+[^+]' | grep -E "${patterns[$i]}" &&
+    ((match++))
+done
+
+for i in "${!email_patterns[@]}"
+do
+    git diff-index -p -M --cached HEAD -- | grep -v "snyk/snyk@"| grep -v "create-release@latest"| grep -v "checkout@master"| grep -v "Publish-Docker-Github-Action@master"| grep -v "docker@master"|
+    grep '^+[^+]' | grep -Eq "${email_patterns[$i]}" &&
+    echo "Blocking commit: ${email_descriptions[$i]} detected in patch" &&
+    echo "This part of your change triggered this issue:" &&
+    git diff-index -p -M --cached HEAD -- | grep -v "snyk/snyk@"| grep -v "create-release@latest"| grep -v "checkout@master"| grep -v "Publish-Docker-Github-Action@master"| grep -v "docker@master"|
+    grep '^+[^+]' | grep -E "${email_patterns[$i]}" &&
+    ((match++))
```

Comment:
this section seems a little awkward to maintain reliably. I think it would be better to declare an `email_exceptions` array and have a nested `for` loop. Something like:
```
for pattern in email_patterns
    for exception in email_exceptions
        if exception not in line added && pattern found in line added
        then display warning reason
        and increment match count
```

Reply:
Not sure that will work, because it only iterates one item in the array at a time.  Meaning the first looped exception will trigger as a fail when checking the second exception.
I originally tried the commands in an array to make it neater, but the for loop injected chars around the grep commands, breaking them.

I'm now looking at a case evaluation of the entire array against each pattern.

Reply:
It can only work like:

```
for pattern in patterns
        if all_exceptions not in line added && pattern found in line added ; then
            display warning reason
            and increment match count
```  

Which is how it currently functions, with `all_exceptions` being verbose.
I'm open to suggestion, but atm this works, and I'm happy to see it improved it later.

Reply:
I would really like this to be merged as-is too, and improved incrementally as and when folks have time. At the moment, when folks run into this thing blocking their commits it can take some time to figure out whether or not it's a false positive. @dan-hill2802 any chance you could give your +1 on it please? Thanks.

Reply:
I don't really have the time to improve on this right now. I did test this in a few different ways last month and it is definitely an improvement on the current implementation, so better to merge now and improve later

Reply:
Much appreciated @dan-hill2802